lib/tree.js

```diff
@@ -12,47 +12,132 @@ class Tree {
+            current = current.left
+        } else if(current.right != null) {
+            current = current.right
+        } 
+        return this.find_helper(current, key)
+    }
+
+    // Time Complexity: O(n) since tree not guaranteed balanced, might have to visit all nodes
+    // Space Complexity: O(1)
     find(key) {
-        throw new Error("This method hasn't been implemented yet!");
+        return this.find_helper(this.root, key)
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
+    inorder_helper(current, visited) {
+        if (current.left != null) { //keep checking left
+            this.inorder_helper(current.left, visited)
+        } 
+        visited.push({key: current.key, value: current.value}) 
+        if (current.right != null) {
+            this.inorder_helper(current.right, visited)
+        }
+        return visited
+    }
+
+    // Time Complexity: O(n)
+    // Space Complexity: O(n)
     inorder() {
-        throw new Error("This method hasn't been implemented yet!");
+        if (this.root == null) return []
+        return this.inorder_helper(this.root, [])
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
+    preorder_helper(current, visited) {
+        visited.push({key: current.key, value: current.value}) 
+        if (current.left != null) { //keep checking left
+            this.preorder_helper(current.left, visited)
+        } 
+        if (current.right != null) {
+            this.preorder_helper(current.right, visited)
+        }
+        return visited
+    }
+    // Time Complexity: O(n)
+    // Space Complexity: O(n)
     preorder() {
-        throw new Error("This method hasn't been implemented yet!");
+        if (this.root == null) return []
+        return this.preorder_helper(this.root, [])
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
+    postorder_helper(current, visited) {
+        if (current.left != null) { //keep checking left
+            this.postorder_helper(current.left, visited)
+        } 
+        if (current.right != null) {
+            this.postorder_helper(current.right, visited)
+        }
+        visited.push({key: current.key, value: current.value}) 
+        return visited
+    }
+
+    // Time Complexity: O(n)
+    // Space Complexity: O(n)
     postorder() {
-        throw new Error("This method hasn't been implemented yet!");
+        if (this.root == null) return []
+        return this.postorder_helper(this.root, [])
     }
 
-    // Time Complexity: ?
-    // Space Complexity: ?
+    height_helper(current, height) {
+        if (current.left != null) {
+            height = height + this.height_helper(current.left, height)
+        } else if (current.right != null) {
+            height = height + this.height_helper(current.right, height)
+        } 
+        return height
+    }
+
+    // Time Complexity: O(n)
+    // Space Complexity: O(n)
     height() {
```

Comment:
👍  The height space complexity is O(log n) in space complexity if the tree is balanced and O(n) if it's unbalanced.